### PR TITLE
Add cron job for sending nudges

### PR DESCRIPTION
**Pull request by bkkilaru**
Purpose:

- Automated sending of 'goal nudges' from coach number. Patients should receive a random message from a pre-written messages 3 days a week (Tues, Thurs, and Sat) in the afternoon (1PM PT) if they're active. Nudges are added as a 'message template' type.

Trello:

- https://trello.com/c/u4Kea7sx

Tests:

- Integration test for sending nudge messages mirroring integration test for midnight messages. Noted that this set to falsy for both even though the message is being sent.

Related Refactor for `dailyMidnightMessages` in implementing `nudgeMessages`:

- Use lodash (groupBy) to group messages by language so its easier to get all english language messages for example.
- Use lodash (sample) for random message
- nit: `() => dailyMidnightMessages()` can just be `dailyMidnightMessages()`. 
- nit: Moved to `if (!patient.enabled) { return }` to avoid "indenting" to improve readability
- nit: Moved from `if (messages.length < 1) ` to `if (messages.length === 0)` for clarity

Future Improvements:

- Send only if patient has received the first blood glucose reading
- Nudges could be day specific or message content could change depending on the day
- Stratify message sending based on patient need

This is my first PR here. Let me know if you have any feedback! 👋🏽

### CODE REVIEW

File: test/utils/backgroundJobs.integration.test.ts [hunk 1]
```diff
@@ -64,6 +74,16 @@ if (process.env.NODE_ENV === 'development') {
       done();
     });
 
+    it('sends nudge messages', async (done) => {
+      await createPatient('111');
+      await createMessageTemplate();
+      nudgeMessages();
+      await waitJest(400);
```
Comment (g-vega-cl) [inline test/utils/backgroundJobs.integration.test.ts:81]: I think waitJest() is not necessary here. Is it?

File: src/background_jobs/utils.ts [hunk 2]
```diff
@@ -15,30 +16,65 @@ export const dailyMidnightMessages = async () => {
...
+  patients.forEach(async (patient) => {
+    if (!patient.enabled) {
+      return;
     }
+
+    const NUDGES_BY_LANGUAGE = groupBy(NUDGES, 'language');
+    const possibleMessages = NUDGES_BY_LANGUAGE[patient.language];
+    const message = sample(possibleMessages)?.text;
+
+    if (possibleMessages.length === 0) {
```
Comment (g-vega-cl) [inline src/background_jobs/utils.ts:60]: I think this could fail if `NUDGES` is empty. Or if we don't have nudges for a specific language.
 * `possibleMessages` could be undefined which would return the error `TypeError: Cannot read property 'length' of undefined`.
  *  Adding  a "?" to possible messages could allow us to avoid this error. `if (possibleMessages?.length === 0) {`
 * `message` could be undefined and it would break our `Message` validation: `ValidationError: Message validation failed: message: Path `message` is required.`
  * maybe a `try..catch` could help us.

This edge cases are unlikely to happen. But it might be worth checking. What do you think?

**Review by g-vega-cl — COMMENTED** [review]
The code is great! I found a couple edge cases in `utils.ts` that we might want to check out.

File: test/utils/backgroundJobs.integration.test.ts (hunk 1 above)
Comment (rweald) [reply test/utils/backgroundJobs.integration.test.ts:81]: It is since the code is not waiting on `nudgeMessages` which is async. This is a bit of a workaround to try and prevent the test from hanging for a long period of time which would be possible with `await nudgeMessages`

Personally I'd probably just do that await since it's a bit less likely to lead to potential flakey tests, but this does work.

File: src/background_jobs/utils.ts [hunk 3]
```diff
@@ -15,30 +16,65 @@ export const dailyMidnightMessages = async () => {
...
+    const newMessage = new Message({
+      patientID: new ObjectId(patient._id),
+      phoneNumber: patient.phoneNumber,
+      date: new Date(),
+      message,
+      sender: 'BOT',
+      sent: false,
+      isCoachingMessage: true,
+    });
+    await newMessage.save();
```
Comment (rweald) [inline src/background_jobs/utils.ts:77]: If this throws an error I think it will crash the background job for remaining members? Since this is just 1 big loop we probably want to log and then continue?

File: src/background_jobs/utils.ts (hunk 2 above)
Comment (rweald) [reply src/background_jobs/utils.ts:60]: Your instinct is definitely right here. I think the real issue is that you already have a optional lookup on the proceeding line (58) so why not just have: 

```
if(!message) {
 do stuff
}
```

Since things are already in place for const message to be `string | undefined`

**Review by rweald — COMMENTED** [review]
I'd probably suggest we make the 1 change to prevent the nudge function from crashing on 1 save failure and not running for everyone else, but otherwise G2G

File: test/utils/backgroundJobs.integration.test.ts (hunk 1 above)
Comment (bkkilaru) [reply test/utils/backgroundJobs.integration.test.ts:81]: Sounds good on moving to `await`. Will leave this alone in this PR since we're using `waitJest()` across a number of these cron tests in this file. Agree we could refactor this more.

File: src/background_jobs/utils.ts (hunk 2 above)
Comment (bkkilaru) [reply src/background_jobs/utils.ts:60]: Cool. Moving from `if (possibleMessages.length === 0)` to `if(!message)`.

File: src/background_jobs/utils.ts (hunk 3 above)
Comment (bkkilaru) [reply src/background_jobs/utils.ts:77]: Hm, wasn't sure if that would actually terminate the `forEach()` in the event it throws an error. Possibly a `for/of` would be better here. 

Will add a simple `try...catch` for `await newMessage.save()`.
